src/optimizer_interface.jl

```diff
@@ -598,22 +598,70 @@ function optimize!(
     try
         MOI.optimize!(backend(model))
     catch err
-        # TODO: This error also be thrown also in MOI.set() if the solver is
-        # attached. Currently we catch only the more common case. More generally
-        # JuMP is missing a translation layer from MOI errors to JuMP errors.
-        if err isa MOI.UnsupportedAttribute{MOI.NLPBlock}
-            error(
-                "The solver does not support nonlinear problems " *
-                "(that is, NLobjective and NLconstraint).",
-            )
-        else
-            rethrow(err)
-        end
+        _rethrow_moi_error(err)
     end
     model.is_model_dirty = false
     return
 end
 
+# TODO: These errors could also be thrown in MOI.add_constraint or MOI.set() if
+# the solver is attached. Currently we catch only the more common case. More
+# generally JuMP is missing a translation layer from MOI errors to JuMP errors.
+_rethrow_moi_error(err) = rethrow(err)
+
+function _rethrow_moi_error(::MOI.UnsupportedAttribute{MOI.NLPBlock})
+    return error(
+        "The solver does not support nonlinear problems " *
+        "(that is, NLobjective and NLconstraint).",
```

Comment:
So it does support it if you use `@constraint` ? The message is unclear

Reply:
NLPBlock in the context of JuMP comes from the `@NL` macros. This was the old error message. I didn't change it. It's also pretty irrelevant now so I don't see much need to change.